### ISSUE TITLE
a problem with druid jdbc support

### ISSUE THREAD

**Issue by shizhengchao** [apache/druid#7931]
when i use `Avatica JDBC driver` connected to `druid router`，i got an error:
```
org.apache.druid.java.util.common.IAE: Received an Avatica request without a connectionId.
```
The following is a detailed description：
### Affected Version

0.14

### Description
when i use `Avatica JDBC driver` connected to `druid router`，when `avatica-remote` send `execute` request to `druid router`, an error happend , i found the `connectionId` is under the key `statementHandle`  like this: 
```json
{"request":"execute","statementHandle":{"connectionId":"b1511e18-c6b8-4c79-a166-977c204511ef","id":1,"signature":{"columns":[{"ordinal":0,"autoIncrement":false,"caseSensitive":true,"searchable":false,"currency":false,"nullable":0,"signed":true,"displaySize":3,"label":"__time","columnName":"__time","schemaName":null,"precision":3,"scale":-2147483648,"tableName":null,"catalogName":null,"type":{"type":"scalar","id":93,"name":"TIMESTAMP","rep":"NUMBER"},"readOnly":true,"writable":false,"definitelyWritable":false,"columnClassName":"java.sql.Timestamp"},{"ordinal":1,"autoIncrement":false,"caseSensitive":true,"searchable":false,"currency":false,"nullable":0,"signed":true,"displaySize":19,"label":"added","columnName":"added","schemaName":null,"precision":19,"scale":0,"tableName":null,"catalogName":null,"type":{"type":"scalar","id":-5,"name":"BIGINT","rep":"NUMBER"},"readOnly":true,"writable":false,"definitelyWritable":false,"columnClassName":"java.lang.Long"},{"ordinal":2,"autoIncrement":false,"caseSensitive":true,"searchable":false,"currency":false,"nullable":1,"signed":true,"displaySize":-1,"label":"channel","columnName":"channel","schemaName":null,"precision":-1,"scale":-2147483648,"tableName":null,"catalogName":null,"type":{"type":"scalar","id":12,"name":"VARCHAR","rep":"STRING"},"readOnly":true,"writable":false,"definitelyWritable":false,"columnClassName":"java.lang.String"},{"ordinal":3,"autoIncrement":false,"caseSensitive":true,"searchable":false,"currency":false,"nullable":1,"signed":true,"displaySize":-1,"label":"cityName","columnName":"cityName","schemaName":null,"precision":-1,"scale":-2147483648,"tableName":null,"catalogName":null,"type":{"type":"scalar","id":12,"name":"VARCHAR","rep":"STRING"},"readOnly":true,"writable":false,"definitelyWritable":false,"columnClassName":"java.lang.String"},{"ordinal":4,"autoIncrement":false,"caseSensitive":true,"searchable":false,"currency":false,"nullable":1,"signed":true,"displaySize":-1,"label":"comment","columnName":"comment","schemaName":null,"precision":-1,"scale":-2147483648,"tableName":null,"catalogName":null,"type":{"type":"scalar","id":12,"name":"VARCHAR","rep":"STRING"},"readOnly":true,"writable":false,"definitelyWritable":false,"columnClassName":"java.lang.String"},{"ordinal":5,"autoIncrement":false,"caseSensitive":true,"searchable":false,"currency":false,"nullable":1,"signed":true,"displaySize":-1,"label":"countryIsoCode","columnName":"countryIsoCode","schemaName":null,"precision":-1,"scale":-2147483648,"tableName":null,"catalogName":null,"type":{"type":"scalar","id":12,"name":"VARCHAR","rep":"STRING"},"readOnly":true,"writable":false,"definitelyWritable":false,"columnClassName":"java.lang.String"},{"ordinal":6,"autoIncrement":false,"caseSensitive":true,"searchable":false,"currency":false,"nullable":1,"signed":true,"displaySize":-1,"label":"countryName","columnName":"countryName","schemaName":null,"precision":-1,"scale":-2147483648,"tableName":null,"catalogName":null,"type":{"type":"scalar","id":12,"name":"VARCHAR","rep":"STRING"},"readOnly":true,"writable":false,"definitelyWritable":false,"columnClassName":"java.lang.String"},{"ordinal":7,"autoIncrement":false,"caseSensitive":true,"searchable":false,"currency":false,"nullable":0,"signed":true,"displaySize":19,"label":"deleted","columnName":"deleted","schemaName":null,"precision":19,"scale":0,"tableName":null,"catalogName":null,"type":{"type":"scalar","id":-5,"name":"BIGINT","rep":"NUMBER"},"readOnly":true,"writable":false,"definitelyWritable":false,"columnClassName":"java.lang.Long"},{"ordinal":8,"autoIncrement":false,"caseSensitive":true,"searchable":false,"currency":false,"nullable":0,"signed":true,"displaySize":19,"label":"delta","columnName":"delta","schemaName":null,"precision":19,"scale":0,"tableName":null,"catalogName":null,"type":{"type":"scalar","id":-5,"name":"BIGINT","rep":"NUMBER"},"readOnly":true,"writable":false,"definitelyWritable":false,"columnClassName":"java.lang.Long"},{"ordinal":9,"autoIncrement":false,"caseSensitive":true,"searchable":false,"currency":false,"nullable":1,"signed":true,"displaySize":-1,"label":"isAnonymous","columnName":"isAnonymous","schemaName":null,"precision":-1,"scale":-2147483648,"tableName":null,"catalogName":null,"type":{"type":"scalar","id":12,"name":"VARCHAR","rep":"STRING"},"readOnly":true,"writable":false,"definitelyWritable":false,"columnClassName":"java.lang.String"},{"ordinal":10,"autoIncrement":false,"caseSensitive":true,"searchable":false,"currency":false,"nullable":1,"signed":true,"displaySize":-1,"label":"isMinor","columnName":"isMinor","schemaName":null,"precision":-1,"scale":-2147483648,"tableName":null,"catalogName":null,"type":{"type":"scalar","id":12,"name":"VARCHAR","rep":"STRING"},"readOnly":true,"writable":false,"definitelyWritable":false,"columnClassName":"java.lang.String"},{"ordinal":11,"autoIncrement":false,"caseSensitive":true,"searchable":false,"currency":false,"nullable":1,"signed":true,"displaySize":-1,"label":"isNew","columnName":"isNew","schemaName":null,"precision":-1,"scale":-2147483648,"tableName":null,"catalogName":null,"type":{"type":"scalar","id":12,"name":"VARCHAR","rep":"STRING"},"readOnly":true,"writable":false,"definitelyWritable":false,"columnClassName":"java.lang.String"},{"ordinal":12,"autoIncrement":false,"caseSensitive":true,"searchable":false,"currency":false,"nullable":1,"signed":true,"displaySize":-1,"label":"isRobot","columnName":"isRobot","schemaName":null,"precision":-1,"scale":-2147483648,"tableName":null,"catalogName":null,"type":{"type":"scalar","id":12,"name":"VARCHAR","rep":"STRING"},"readOnly":true,"writable":false,"definitelyWritable":false,"columnClassName":"java.lang.String"},{"ordinal":13,"autoIncrement":false,"caseSensitive":true,"searchable":false,"currency":false,"nullable":1,"signed":true,"displaySize":-1,"label":"isUnpatrolled","columnName":"isUnpatrolled","schemaName":null,"precision":-1,"scale":-2147483648,"tableName":null,"catalogName":null,"type":{"type":"scalar","id":12,"name":"VARCHAR","rep":"STRING"},"readOnly":true,"writable":false,"definitelyWritable":false,"columnClassName":"java.lang.String"},{"ordinal":14,"autoIncrement":false,"caseSensitive":true,"searchable":false,"currency":false,"nullable":1,"signed":true,"displaySize":-1,"label":"metroCode","columnName":"metroCode","schemaName":null,"precision":-1,"scale":-2147483648,"tableName":null,"catalogName":null,"type":{"type":"scalar","id":12,"name":"VARCHAR","rep":"STRING"},"readOnly":true,"writable":false,"definitelyWritable":false,"columnClassName":"java.lang.String"},{"ordinal":15,"autoIncrement":false,"caseSensitive":true,"searchable":false,"currency":false,"nullable":1,"signed":true,"displaySize":-1,"label":"namespace","columnName":"namespace","schemaName":null,"precision":-1,"scale":-2147483648,"tableName":null,"catalogName":null,"type":{"type":"scalar","id":12,"name":"VARCHAR","rep":"STRING"},"readOnly":true,"writable":false,"definitelyWritable":false,"columnClassName":"java.lang.String"},{"ordinal":16,"autoIncrement":false,"caseSensitive":true,"searchable":false,"currency":false,"nullable":1,"signed":true,"displaySize":-1,"label":"page","columnName":"page","schemaName":null,"precision":-1,"scale":-2147483648,"tableName":null,"catalogName":null,"type":{"type":"scalar","id":12,"name":"VARCHAR","rep":"STRING"},"readOnly":true,"writable":false,"definitelyWritable":false,"columnClassName":"java.lang.String"},{"ordinal":17,"autoIncrement":false,"caseSensitive":true,"searchable":false,"currency":false,"nullable":1,"signed":true,"displaySize":-1,"label":"regionIsoCode","columnName":"regionIsoCode","schemaName":null,"precision":-1,"scale":-2147483648,"tableName":null,"catalogName":null,"type":{"type":"scalar","id":12,"name":"VARCHAR","rep":"STRING"},"readOnly":true,"writable":false,"definitelyWritable":false,"columnClassName":"java.lang.String"},{"ordinal":18,"autoIncrement":false,"caseSensitive":true,"searchable":false,"currency":false,"nullable":1,"signed":true,"displaySize":-1,"label":"regionName","columnName":"regionName","schemaName":null,"precision":-1,"scale":-2147483648,"tableName":null,"catalogName":null,"type":{"type":"scalar","id":12,"name":"VARCHAR","rep":"STRING"},"readOnly":true,"writable":false,"definitelyWritable":false,"columnClassName":"java.lang.String"},{"ordinal":19,"autoIncrement":false,"caseSensitive":true,"searchable":false,"currency":false,"nullable":1,"signed":true,"displaySize":-1,"label":"user","columnName":"user","schemaName":null,"precision":-1,"scale":-2147483648,"tableName":null,"catalogName":null,"type":{"type":"scalar","id":12,"name":"VARCHAR","rep":"STRING"},"readOnly":true,"writable":false,"definitelyWritable":false,"columnClassName":"java.lang.String"}],"sql":"select * from wikipedia where comment = 'added project'","parameters":[],"cursorFactory":{"style":"ARRAY","clazz":null,"fieldNames":null},"statementType":"SELECT"}},"parameterValues":[],"maxRowCount":100}
```
so, an error occured in `org.apache.druid.server.AsyncQueryForwardingServlet#service`:
```java
if (isAvatica) {
     .....
     String connectionId = getAvaticaConnectionId(requestMap);
      ......
}
private static String getAvaticaConnectionId(Map<String, Object> requestMap)
  {
   // There should be an error here, because the connectionId is in `statementHandle`
    Object connectionIdObj = requestMap.get("connectionId");
    if (connectionIdObj == null) {
      throw new IAE("Received an Avatica request without a connectionId.");
    }
    if (!(connectionIdObj instanceof String)) {
      throw new IAE("Received an Avatica request with a non-String connectionId.");
    }

    return (String) connectionIdObj;
  }
```

**Comment by shizhengchao** [comment]
So this method `getAvaticaConnectionId`  is not compatible with all requests from avatica-remote

**Comment by shizhengchao** [comment]
@gianm

**Comment by fz12805** [comment]
I got the same problem, when executing with jdbc connection, the request body is like
`{"request":"prepareAndExecute","connectionId":"bc4d1171-60d9-409b-b350-8de8cf205efd","statementId":1,"sql":"SELECT dim4, SUM(metric1Sum)/COUNT(*) as va FROM apm_common_v1 WHERE \"__time\" > TIMESTAMP '2019-10-01 16:18:31' AND \"__time\" < TIMESTAMP '2019-10-25 17:18:31' AND dim2 = '1' AND dim3 = 'android'  AND dim1 in ('network') group by dim4 order by va desc limit 10","maxRowsInFirstFrame":-1,"maxRowCount":-1}`

but when executing use mybastis, the request is like above with connectionId in the body of statementHandle, which can't be parsed by server

**Comment by fz12805** [comment]
We come around this problem by setting @Options(statementType = StatementType.STATEMENT) on MyBatis Mapper, which force MyBatis to use Statement instead of PreparedStatement

**Comment by gianm** [comment]
Oh, interesting, we should fix this in the Router.

Does it work when you query directly against the Broker?

**Comment by gianm** [comment]
By the way, a patch against AsyncQueryForwardingServlet's `getAvaticaConnectionId` method to fix this would be greatly appreciated.

**Comment by shizhengchao** [comment]
> Oh, interesting, we should fix this in the Router.
> 
> Does it work when you query directly against the Broker?

it work when i query directly against the Broker.

**Comment by Doslin** [comment]
> > Oh, interesting, we should fix this in the Router.
> > Does it work when you query directly against the Broker?
> 
> it work when i query directly against the Broker.
How to solve this problem , does it work properly in router mode?
@shizhengchao

**Comment by dars1608** [comment]
Are there any plans to patch this in the next updates?

**Comment by gianm** [comment]
Fixed by #10272.
